### PR TITLE
Fix embind benchmark.

### PR DESCRIPTION
**Pull request by brendandahl**
- Fix compile errors related to memory view.
- Update exported functions, so it runs correctly.
- Uncomment other benchmarks.

### CODE REVIEW

**Review by kripken — APPROVED** [review]
I didn't remember we even had an embind benchmark... :smile: it's not hooked up to `test_benchmark.py`. But I guess it can't be because that compares to native builds.

How useful is the benchmark, do you think?

**Comment by brendandahl** [comment]
I find some of the basic benchmarks interesting, since they allow you to evaluate overhead of embind. They have things implemented three ways: 
1) pure c++
2) js->wasm (direct calls, no binding layer)
3) embind

I'm not sure how useful the various interface benchmarks are. I haven't really seen anyone use embind that way, but I'll also haven't seen much embind in the wild yet.

If anything the file seems like a good starting place to add some benchmarks (such as in #17292).

File: tests/embind/build_benchmark [hunk 1]
```diff
@@ -1,2 +1,2 @@
 #!/bin/bash
-../../emcc --minify 0 --bind --post-js embind.benchmark.js -O2 --shell-file shell.html -o embind_benchmark.html embind_benchmark.cpp
+../../emcc -sEXPORTED_FUNCTIONS=_main,_get_counter,_increment_counter,_returns_input,_sum_int,_sum_float --minify 0 --bind --js-library=embind.benchmark.js -O2 --shell-file shell.html -o embind_benchmark.html embind_benchmark.cpp
```
Comment (sbc100) [inline tests/embind/build_benchmark:2]: Was converting this file from a post-js to a js-library necessary? 

Perhaps its was the case that providing externs to C/C++ was possible via a post-js in previous version of emscripten?  I wonder if `-sWARN_ON_UNDEFINED_FUNCTIONS=0` would have allowed the post-js method to continue to work?  Otherwise, how did it ever work?

File: tests/embind/embind_benchmark.cpp [hunk 2]
```diff
@@ -149,7 +149,7 @@ class Interface
     virtual void call0() = 0;
     virtual std::wstring call1(const std::wstring& str1, const std::wstring& str2) = 0;
     virtual void call_with_typed_array(size_t size, const void*) = 0;
-    virtual void call_with_memory_view(size_t size, const void*) = 0;
+    virtual void call_with_memory_view(size_t size, const unsigned int*) = 0;
```
Comment (sbc100) [inline tests/embind/embind_benchmark.cpp:152]: Why this change?

File: tests/embind/build_benchmark (hunk 1 above)
Comment (brendandahl) [reply tests/embind/build_benchmark:2]: I was wondering how this use to work. Is one way or the other preferred?

File: tests/embind/embind_benchmark.cpp (hunk 2 above)
Comment (brendandahl) [reply tests/embind/embind_benchmark.cpp:152]: I was getting the compile error
```
emscripten/wire.h:402:18: error: invalid application of 'sizeof' to an incomplete type 'void'
```
I'm a bit confused how this ever worked.

Comment (sbc100) [reply tests/embind/embind_benchmark.cpp:152]: But the line above was fine?  Strange.

File: tests/embind/build_benchmark (hunk 1 above)
Comment (brendandahl) [reply tests/embind/build_benchmark]: `-sWARN_ON_UNDEFINED_SYMBOLS=0` works, do you want me to revert these changes?

Comment (sbc100) [reply tests/embind/build_benchmark]: No, its better to use `--js-library` file I think.

File: tests/embind/embind_benchmark.cpp (hunk 2 above)
Comment (brendandahl) [reply tests/embind/embind_benchmark.cpp:152]: Yeah, `sizeof(void)` errors, but `sizeof(int)` is okay. `sizeof(void*)` is also okay, but not applied in this case.

**Comment by xbcnn** [comment]
@brendandahl 
`tests/embind/build_benchmark` seems a bash script which invokes `emcc` to compile.  I am wondering whether it uses the right header files, e.g., the `emscripten/val.h`, from the repo instead of the one from installed emsdk/.  Any tricks?

**Comment by brendandahl** [comment]
> @brendandahl `tests/embind/build_benchmark` seems a bash script which invokes `emcc` to compile. I am wondering whether it uses the right header files, e.g., the `emscripten/val.h`, from the repo instead of the one from installed emsdk/. Any tricks?

It should be using the correct val.h, but I'm guessing it's picking up the cached version. Try running `../.././embuilder build sysroot --force` or `../.././emcc --clear-cache`.

**Comment by xbcnn** [comment]
> Try running `../.././embuilder build sysroot --force` or `../.././emcc --clear-cache`.

OK.
